src/devices/cpu/z180/z180asci.cpp

```diff
@@ -0,0 +1,402 @@
+}
+
+void z180asci_channel_base::device_resolve_objects()
+{
+	// resolve callbacks
+	m_txa_handler.resolve_safe();
+	m_rts_handler.resolve_safe();
+}
+
+void z180asci_channel_base::device_start()
+{
+	save_item(NAME(m_asci_cntla));
+	save_item(NAME(m_asci_cntlb));
+	save_item(NAME(m_asci_stat));
+	save_item(NAME(m_asci_tdr));
+	save_item(NAME(m_asci_rdr));
+	if (m_ext)
+	{
+		save_item(NAME(m_asci_ext));
+		save_item(NAME(m_asci_tc.w));
+	}
+	save_item(NAME(m_divisor));
+	transmit_register_reset();
+	receive_register_reset();
+}
+
+void z180asci_channel_base::device_reset()
+{
+	m_asci_cntla = 0;
+	m_asci_cntlb = 7; // SS2,SS1 and SS0 are 1 after reset
+	m_asci_tdr = 0;
+	m_asci_rdr = 0;
+	m_asci_stat = 2; // set TDRE=1 on reset
+	m_irq = 0;
+	m_cts = 0;
+	m_dcd = 0;
+	m_asci_ext = 0;
+	m_asci_tc.w = 0;
+	m_divisor = 1;
+}
+
+void z180asci_channel_base::device_clock_changed()
+{
+	LOG("Z180 ASCI%d set bitrate %d\n", m_id, uint32_t(clock() / m_divisor));
+	set_tra_rate(clock(), m_divisor);
+	set_rcv_rate(clock(), m_divisor);
+}
+
+uint8_t z180asci_channel_base::cntla_r()
+{
+	uint8_t data = m_asci_cntla;
+	LOG("Z180 CNTLA%d rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel_base::cntlb_r()
+{
+	uint8_t data = m_asci_cntlb;
+	LOG("Z180 CNTLB%d rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel_base::stat_r()
+{
+	uint8_t data = m_asci_stat;
+	LOG("Z180 STAT%d  rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel_base::tdr_r()
+{
+	uint8_t data = m_asci_tdr;
+	LOG("Z180 TDR%d   rd $%02x\n", m_id, data);
+	return data;
+}
+
+uint8_t z180asci_channel_base::rdr_r()
+{
+	uint8_t data = m_asci_rdr;
+	LOG("Z180 RDR%d   rd $%02x\n", m_id, data);
+	m_asci_stat &= ~0x80;
+
+	return data;
+}
+
+uint8_t z180asci_channel_base::asext_r()
+{
+	LOG("Z180 ASEXT%d rd $%02x\n", m_id, m_asci_ext);
+	return m_asci_ext;
+}
+
+uint8_t z180asci_channel_base::astcl_r()
+{
+	uint8_t data = m_asci_tc.b.l;
+	LOG("Z180 ASTC%dL rd $%02x ($%04x)\n", m_id, data, m_asci_tc.w);
+	return data;
+}
+
+uint8_t z180asci_channel_base::astch_r()
+{
+	uint8_t data = m_asci_tc.b.h;
+	LOG("Z180 ASTC%dH rd $%02x ($%04x)\n", m_id, data, m_asci_tc.w);
+	return data;
+}
+
+void z180asci_channel_base::cntla_w(uint8_t data)
+{
+	LOG("Z180 CNTLA%d wr $%02x\n", m_id, data);
+	m_asci_cntla = data;
+	set_data_frame(1, (data & 4) ? 8 : 7, (data & 2) ? PARITY_ODD : PARITY_NONE, (data & 2) ? STOP_BITS_2 : STOP_BITS_1);
+}
+
+void z180asci_channel_base::cntlb_w(uint8_t data)
+{
+	LOG("Z180 CNTLB%d wr $%02x\n", m_id, data);
+	m_asci_cntlb = data;
+
+	m_divisor = 1<<(m_asci_cntlb & 0x07);
+	m_divisor *= ((m_asci_cntlb & 0x20) ? 30 : 10);
+	m_divisor *= ((m_asci_cntlb & 0x08) ? 64 : 16);
+
+	device_clock_changed();
+}
+
+void z180asci_channel_base::tdr_w(uint8_t data)
+{
+	LOG("Z180 TDR%d   wr $%02x\n", m_id, data);
+	m_asci_tdr = data;
+	m_asci_stat &= ~0x02; // clear TDRE
+	transmit_byte(m_asci_tdr);
+}
```

Comment:
How are you going to handle the MPB and MP flags for software control of the trailing bit when you’re using diserial?